arangod/Utils/SupportInfoBuilder.cpp

```diff
@@ -0,0 +1,762 @@
+#include "Utils/SingleCollectionTransaction.h"
+#include "VocBase/LogicalCollection.h"
+#include "VocBase/LogicalView.h"
+#include "VocBase/Methods/Collections.h"
+#include "VocBase/Methods/Databases.h"
+#include "VocBase/Methods/Indexes.h"
+
+#include <velocypack/Builder.h>
+#include <absl/strings/str_replace.h>
+
+#include "Logger/LogMacros.h"
+
+using namespace arangodb;
+using namespace arangodb::rest;
+using namespace std::literals;
+
+namespace {
+network::Headers buildHeaders() {
+  auto auth = AuthenticationFeature::instance();
+
+  network::Headers headers;
+  if (auth != nullptr && auth->isActive()) {
+    headers.try_emplace(StaticStrings::Authorization,
+                        "bearer " + auth->tokenCache().jwtToken());
+  }
+  return headers;
+}
+}  // namespace
+
+void SupportInfoBuilder::addDatabaseInfo(VPackBuilder& result,
+                                         VPackSlice infoSlice,
+                                         ArangodServer& server) {
+  DatabaseFeature& dbFeature = server.getFeature<DatabaseFeature>();
+
+  std::vector<std::string> databases = methods::Databases::list(server, "");
+
+  containers::FlatHashMap<std::string_view, uint32_t> dbViews;
+  for (auto const& database : databases) {
+    auto vocbase = dbFeature.lookupDatabase(database);
+
+    LogicalView::enumerate(*vocbase,
+                           [&dbViews, &database = std::as_const(database)](
+                               LogicalView::ptr const& view) -> bool {
+                             dbViews[database]++;
+
+                             return true;
+                           });
+  }
+
+  struct DbCollStats {
+    size_t numDocColls;
+    size_t numGraphColls;
+    size_t numSmartColls;
+    size_t numDisjointGraphs;
+    VPackBuilder builder;
+  };
+
+  std::unordered_map<std::string_view, DbCollStats> visitedDatabases;
+  containers::FlatHashMap<size_t, containers::FlatHashSet<std::string_view>>
+      visitedColls;
+  containers::FlatHashMap<size_t, size_t> collNumDocs;
+  // merge all collections belonging to the database into
+  // the same object, as the database might be in more
+  // than one dbserver
+  for (auto const dbItFromServers : velocypack::ObjectIterator(infoSlice)) {
+    for (auto const dbIt : velocypack::ArrayIterator(dbItFromServers.value)) {
+      std::string_view dbName = dbIt.get("name").stringView();
+      if (visitedDatabases.find(dbName) == visitedDatabases.end()) {
+        visitedDatabases.insert(std::make_pair(dbName, DbCollStats()));
+        visitedDatabases[dbName].builder.openObject();
+        visitedDatabases[dbName].builder.add("n_views",
+                                             VPackValue(dbViews[dbName]));
+        visitedDatabases[dbName].builder.add(
+            "single_shard", VPackValue(dbIt.get("single_shard").getBoolean()));
+        visitedDatabases[dbName].builder.add("colls",
+                                             VPackValue(VPackValueType::Array));
```

Comment:
This repeatedly looks up the `dbName` key in `visitedDatabases`. Works, but we just inserted to key, so we don't need to look it up repeatedly.